File: 01_languages/clojure/destructuring_and_pattern_matching.clj
(def a-list [1 2 3 4 5])

(println a-list)

; Destructuring a list
; Tuple can be represented using a List
(let [[first second] a-list] 
  (println [first, second]))
     
(let [[first second & rest] a-list]
  (println [first, second, rest]))


; Destructuring a map (Associative Destructuring)
(def a-name {:first "Dhaval" :last "Dalal" :salutation "Mr."})

(println a-name)

(let [{fname :first lname :last salutation :salutation} a-name]
  (println fname lname))


(defn capitalize [{fname :first lname :last}] 
  (str (.toUpperCase fname) " " (.toUpperCase lname)))

(println (capitalize a-name))

; Idiomatic Clojure does not have pattern matching like that in Scala, 
; but we can use multi-methods to achieve that.
(defmulti distance
  ; Dispatcher Function
  ; Out-of-box Clojure has case - but it does exact match and
  ; hence resorted to cond.  Other option is to use core.match
  ; library (not a part of standard distribution).
  (fn [x y] 
    (cond (= x y [0 0]) [:zero :zero]
          (= x [0 0])   [:zero :any]
          (= y [0 0])   [:any :zero]
          :else [x y])))
          
(defn dist [x y] (Math/sqrt (+ (* x x) (* y y))))   
(defmethod distance [:zero :zero] [x y] 0)
(defmethod distance [:any :zero]  [[x1 y1] y] (dist x1 y1))
(defmethod distance [:zero :any]  [x [x2 y2]] (dist x2 y2))
(defmethod distance :default [[x1 y1] [x2 y2]] (dist (- x2 x1) (- y2 y1)))

(println (distance [0 0] [0 0]))
(println (distance [3 0] [0 0]))
(println (distance [0 0] [0 4]))
(println (distance [3 0] [0 4]))

File: 01_languages/scala/destructuring_and_pattern_matching.scala
// In FP, Destructuring is about extracting data from the innards of any structure.

// Destructuring list
val list = 1::2::3::4::5::6::Nil
// Destructuring gives us a short-hand way of naming parts of the data-structure.
val (first::second::rest) = list
println(first)
println(second)
println(rest)

// Destructuring nTuples
val (x, y) = (20, 30)
println(x)
println(y)

val (p, q, r) = (20, 30, 40)
println(p)
println(q)
println(r)

case class Name(fName: String, lName: String, salutation: String)

// Pattern Matching
def capitalize(name: Name) = name match {
  case Name(firstName, lastName, _) => firstName.toUpperCase + " " + lastName.toUpperCase
}

println(capitalize(Name("Dhaval", "Dalal", "Mr.")))

type Point = (Double, Double)
// Pattern Matching
def distance(p1: Point, p2: Point): Double = {
  def pythagorean(x: Double, y: Double): Double = Math.sqrt(x * x + y * y)
  (p1, p2) match {
    case (p1, p2) if p1 == p2 => 0
    case (p1, p2) if p1 == (0,0) || p2 == (0,0) => 
      val (x,y) = if (p1 == (0,0)) p2 else p1 
      pythagorean(x, y)
    case ((x1, y1), (x2, y2)) => pythagorean(x2 - x1, y2 - y1)
  }
}

println(distance((0,0), (0,0)))
println(distance((3,0), (0,0)))
println(distance((0,0), (0,4)))
println(distance((3,0), (0,4)))
